docs/design/2022-07-20-insert-support.md

```diff
@@ -0,0 +1,112 @@
+- [References](#references)
+
+## Introduction
+
+New feature: TiSpark Insert SQL Feature
+
+## Motivation or Background
+
+TiSpark is a connector that provides read, write and delete support.
+But it only supports writing data using DataSource API, users can't use Spark SQL to write data.
+
+It's common for users to use SQL statements in the big-data workflow. We are going to support TiSpark to
+write using Spark SQL.
+
+## Goals
+**Must have**
+- Support write data by Spark SQL
+- Compatible with Spark 3.x
+
+**Nice to have**
+- Type convert compatible with the previous version
+- User doc
+
+## API
+Use spark SQL to write data in TiSpark.
+```
+// insert with spark sql
+spark.sql("insert into tidb_db.db.table values(x, x...)(y, y...)...")
+```
+
+## Detailed Design
+Here are the main steps:
+- Add new capability
+- remove write options
+- Implement V2 interface using V1 code
+- Compatible with different Spark versions
+
+### Add new capability
+Spark will check the capabilities of TiDBTable. If we don't add corresponding capability, Spark will throw an exception.
+
+### Remove write options
+Different from writing using DataSource API, Spark SQL can't get required information like tidb address and username.
+We use it to create a JDBC client to connect with TiDB, but we only need to use the JDBC client to update statics now.
+
+So we are going to remove the options check before writing and useless JDBC client code.
+
+### Implement V2 interface using V1 code
+Spark DataSource V2 is evolving and not suitable for TiSpark write model currently.
+Spark provides us with an interface that is implemented by DataSource V1 and leverages the DataSource V2 write code path.
+
+We can use this interface to implement INSERT SQL without rewriting all write codes.
+
+### Compatible with different Spark versions
+The interface we used to combine DataSource V1 and DataSource V2 is different between
+Spark 3.0, 3.1, and 3.2. So we need to write code for each version. Then we can use reflection to get object of
+corresponding version。
+
+To be specific, we will use V1WriteBuilder in 3.0 and 3.1, and use V1Write in 3.2.
+
+### Why not V2 write model
+There is a deduplicate operation in TiSpark that will handle the problem that data have the same unique key.
+DataSource V2 will split data into different executors, and we can't control it. This may lead to different
+executors writing data with the same unique key and cause data corruption.
+
+Another problem is about region split. Currently, we split regions to enable parallel write, each executor will write data to only one
+region. Since we can't control how to split data in DataSource V2, each executor may receive data that
+should be written to different regions. If we don't do region split, it may cause performance issue. If we do, region split will be
+done in every executor which also causes performance issue.
+
+
+## Compatibility
+- TiDB support: 4.x and 5.x
+- Spark support: >= 3.0
```

Comment:
We need to support TiDB 6.x. And for Spark, at this time, we support 3.0, 3.1, 3.2 but not 3.3. So I think >= 3.0 is not accurate.